include/modem/pdn.h

```diff
@@ -0,0 +1,196 @@
+#include <stddef.h>
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** @brief PDN family */
+enum pdn_fam {
+	PDN_FAM_IPV4   = 0,
+	PDN_FAM_IPV6   = 1,
+	PDN_FAM_IPV4V6 = 2,
+	PDN_FAM_NONIP  = 3,
+};
+
+/** @brief Additional PDP Context configuration options */
+struct pdn_pdp_opt {
+	/**
+	 * @brief IPv4 address allocation.
+	 * 0 – IPv4 address via Non-access Stratum (NAS) signaling (default)
+	 * 1 – IPv4 address via Dynamic Host Configuration Protocol (DHCP)
+	 */
+	uint8_t ip4_addr_alloc;
+	/**
+	 * @brief NAS Signalling Low Priority Indication.
+	 * 0 – Use Non-access Stratum (NAS) Signalling Low Priority Indication (NSLPI) value
+	 *     from configuration (default)
+	 * 1 – Use value "Not configured" for NAS Signalling Low Priority Indication
+	 */
+	uint8_t nslpi;
+	/**
+	 * @brief Protected transmission of Protocol Configuration Options (PCO).
+	 * 0 – Protected transmission of PCO is not requested (default)
+	 * 1 – Protected transmission of PCO is requested
+	 */
+	uint8_t secure_pco;
+};
+
+/** @brief PDN library event */
+enum pdn_event {
+	PDN_EVENT_CNEC_ESM,	/** +CNEC ESM error code */
+	PDN_EVENT_ACTIVATED,	/** PDN connection activated */
+	PDN_EVENT_DEACTIVATED,	/** PDN connection deactivated */
+	PDN_EVENT_IPV6_UP,	/** PDN has IPv6 connectivity */
+	PDN_EVENT_IPV6_DOWN,	/** PDN has lost IPv6 connectivity */
+};
+
+/** @brief PDN authentication method */
+enum pdn_auth {
+	PDN_AUTH_NONE = 0,	/** No authentication */
+	PDN_AUTH_PAP  = 1,	/** Password authentication protocol */
+	PDN_AUTH_CHAP = 2,	/** Challenge handshake authentication protocol */
+};
+
+/**
+ * @typedef pdn_event_handler_t
+ *
+ * Event handler for PDN events.
+ * If assigned during PDP context creation, the event handler will receive
+ * status information relative to the Packet Data Network connection,
+ * as reported by the AT notifications +CNEC and +GGEV.
+ *
+ * This handler is executed by the same context that dispatches AT notifications.
+ *
+ * @param cid The PDP context ID.
+ * @param event The event.
+ * @param reason The ESM error reason, if available.
+ */
+typedef void (*pdn_event_handler_t)(uint8_t cid, enum pdn_event event,
+				    int reason);
+
+/**
+ * @brief Initialize the PDN library.
+ *
+ * This library depends on the @ref at_cmd and @ref at_notif libraries,
+ * and they must be both initialized before initializing this library.
+ *
+ * @return int Zero on success or a negative errno otherwise.
+ */
+int pdn_init(void);
+
+/**
+ * @brief Create a Packet Data Protocol (PDP) context.
+ *
+ * If a callback is provided via the @c cb parameter, the library will
+ * generate events from the +CNEC and +GGEV AT notifications to report
+ * state of the Packet Data Network (PDN) connection.
+ *
+ * @param[out] cid The ID of the new PDP context.
+ * @param cb Optional event handler.
+ * @return int Zero on success or a negative errno otherwise.
+ */
+int pdn_cid_create(uint8_t *cid, pdn_event_handler_t cb);
```

Comment:
would pdn_ctx_create() or pdn_context_create() be better than _cid_? And the same naming for the rest as well.

Reply:
Yes, I think that's better; will change.

Reply:
Changed to `pdn_ctx_`